Add hard-coded time-dependent maps to BinaryCompactObject

geoffrey4444 opened the pull request:
## Proposed changes

This PR i) removes the TimeDependence option from BinaryCompactObject, ii) adds an (optional, depending on the value of a flag in the metavariables) hard-coded time-dependent map to BinaryCompactObject that is a composition of an expansion and a rotation about the z axis, except if object A or B is excised, then the map in the "layer 1" blocks for that object get a composition of a size, expansion, and rotation map instead. 

[A previous version of this PR also updated ExportCoordinates to be aware of this optional time-dependence, but that is currently removed here, held back for a future PR, to keep this one as straightforward to review as possible.]

This is a foundation for coding up the time-dependent behavior that compact binaries will have. Future PRs will add other time-dependent mappings, such as shape and generic rotations. The map implemented here is the one added here both for simplicity and because it is suitable for the simplest BBH evolution: equal-mass, non spinning BBH evolutions. 

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
After this PR, it will be possible to specify extra options to BinaryCompactObject to initialize time-dependent maps corresponding to size, expansion, and rotation maps. In order to use the time-dependent version of BinaryCompactObject, you'll need to add to the metavariables a struct named `domain` that conforms to `domain::Protocols::Metavariables`. For ExportCoordinates to be aware of the time dependence, it will have to be updated in a future PR to be aware of this new metavariable.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nikwit commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -297,11 +385,25 @@ Domain<3> BinaryCompactObject::create_domain() const noexcept {
+    using CubicScaleMap = domain::CoordinateMaps::TimeDependent::CubicScale<3>;
+    using CubicScaleMapForComposition =
+        domain::CoordinateMap<Frame::Grid, Frame::Inertial, CubicScaleMap>;
+    std::vector<std::unique_ptr<
```
This would be a lot simpler:
```
const auto block_map = std::make_unique<CubicScaleMapForComposition>(...);
for (size_t block = 0; block < number_of_blocks_; ++block) {
      domain.inject_time_dependent_map_for_block(block, block_map -> get_clone());
}
```

I assume you added it like this so in the future  more maps will be added into the composition map in the vector, some of which may apply only to certain blocks. Is that correct? maybe add a comment until then so its clear.

nikwit commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -333,8 +413,44 @@ class BinaryCompactObject : public DomainCreator<3> {
-      std::unique_ptr<domain::creators::time_dependence::TimeDependence<3>>
-          time_dependence = nullptr,
+      const Options::Context& context = {});
+
```
can you add a small comment to describe what the second constructor does different to the first one, i.e. the time dependent maps.

nikwit commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -91,76 +161,94 @@ BinaryCompactObject::BinaryCompactObject(
+    typename AdditionToObjectBRadialRefinementLevel::type
+        addition_to_object_B_radial_refinement_level,
+    const Options::Context& context)
+    // clang-tidy: trivially copyable
```
why not just copy?

nikwit commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -361,10 +463,35 @@ BinaryCompactObject::initial_refinement_levels() const noexcept {
+
+  // Use a 3rd deriv function of time so that it can be used with a control
+  // system.
+  result[expansion_function_of_time_names_[0]] =
```
I assume these are `a` and `b` of the `CubicScaleMap`? Maybe you could add a comment here and/or explain that in the constructor with the expansion function of time.

wthrowe commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -34,6 +36,77 @@ struct Logical;
+  if (use_logarithmic_map_object_A_ and not excise_interior_A_) {
+    PARSE_ERROR(
+        context,
+        "excise_interior_A must be true if use_logarithmic_map_object_A is "
```
This is a user-facing error message, so it should not describe the problem in terms of internal variable names.  I think just changing to camel case will make them match the input file names, which is fine.  (Same for the next one.)

geoffrey4444 replied:
These messages have already been fixed in another commit that will apply after I rebase this PR on latest develop

geoffrey4444 replied:
I double-checked, and after my last rebase, these messages are in fact fixed to not refer to variable names

wthrowe commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -346,6 +463,35 @@ BinaryCompactObject::initial_refinement_levels() const noexcept {
+  std::unordered_map<std::string,
+                     std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>
+      result{};
+  if (!enable_time_dependence_) {
```
! -> not

wthrowe commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -346,6 +463,35 @@ BinaryCompactObject::initial_refinement_levels() const noexcept {
+
+  const double initial_expiration_time =
+      initial_expiration_delta_t_ ? initial_time_ + *initial_expiration_delta_t_
+                                  : std::numeric_limits<double>::max();
```
`max()` or `infinity()`?  Not necessarily wrong, I was just expecting the latter.

geoffrey4444 replied:
I just wanted the functions of time to never expire but also not throw an FPE, so I thought max() was the way to go. I could test if infinity() also got the same result if you like

geoffrey4444 replied:
I'll change this to infinity(), because I realize that i) comparing a finite number to infinity() should not throw a floating point exception, and ii) doing math on the expiration delta t (which does throw an exception) shouldn't happen, so might as well make sure

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -262,14 +279,85 @@ class BinaryCompactObject : public DomainCreator<3> {
+    using group = ExpansionMap;
+  };
+  /// \brief The names of the functions of times to be added to the added to the
+  /// DataBox.
```
Duplicate "added to the".

geoffrey4444 committed:
Remove BinaryCompactObject TimeDependence

geoffrey4444 committed:
Add time-dependent maps to BinaryCompactObject

geoffrey4444 commented:
@nikwit @wthrowe I rebased on develop, removed the ExportCoordinates stuff (will submit a separate PR for that), and added in all the maps I'm currently using for BBH runs (expansion, rotation about z axis, size map). I also incorporated your feedback above.

I think this is ready for another look when you get a chance. My apologies for the delay in responding to your reviews! (If the tests end up failing (they passed locally for me), I'll push a fixup commit...I'm still waiting to see if they pass.)

geoffrey4444 commented:
I don't understand the one failing test...looks like the compiler crashed while building a file I'm pretty sure I did t touch in this PR?

wthrowe commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -418,6 +432,10 @@ Domain<3> BinaryCompactObject::create_domain() const noexcept {
 
   // Inject the hard-coded time-dependence
   if (enable_time_dependence_) {
+    // Note on frames: Because the relevant maps will all be composed before
+    // they are used, all maps here go from Frame::Grid (the frame after the
+    // final time-independent map is applied) to Frame::Inertial
+    // (the frame after the final time-dependent map is applied).
```
I thought we were using Frame::None for compositions.

geoffrey4444 replied:
Are you saying I should switch Frame::Grid to Frame::None? If so, then what is Frame::Grid for?

wthrowe replied:
OK, it looks like you're right and the manipulation functions do require all the maps to be Grid->Inertial.  That doesn't make any sense mathematically, but I guess we go with it.

NoFrame (not None, I misremembered) is for intermediate unnamed frames, which seems like exactly what you have here.  (A frame after scaling but before rotation.)

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -468,6 +481,43 @@ class BinaryCompactObject : public DomainCreator<3> {
+  };
+
+  /// \brief Initial values for functions of time for size maps for objects A,B.
+  struct InitialSizeMapValues {
+    using type = std::array<double, 2>;
+    static constexpr Options::String help = {
+        "SizeMapA, SizeMapB values at initial time."};
+    using group = SizeMap;
+  };
```
How are these options handled depending on whether or not the centers are excised?

geoffrey4444 replied:
If a center is not excised, then the corresponding option is ignored, since the size map is not applied

wthrowe replied:
Document that somewhere.

geoffrey4444 replied:
Documented as \details in each size map tag's documentation

wthrowe reviewed (changes requested):
Best guess on the CI error is that it ran out of memory.  Probably unrelated to this change, so don't worry about it unless it happens again.

wthrowe reviewed (changes requested):
One documentation request.

wthrowe commented:
Looks good.  That's simple enough that I think you can squash without waiting for a second OK.

geoffrey4444 committed:
Add domain::protocols::Metavariables

nilsvu commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -148,6 +138,89 @@ BinaryCompactObject::BinaryCompactObject(
+    std::unique_ptr<domain::BoundaryConditions::BoundaryCondition>
+        outer_boundary_condition,
+    const Options::Context& context)
+    : object_A_(std::move(object_A)),
```
(optional) you can use a [delegating constructor](https://en.cppreference.com/w/cpp/language/constructor#Delegating_constructor) here: Write `: BinaryCompactObject(std::move(object_A), ...)` to delegate to the non-time-dependent constructor, then init the time-dependent member vars in the body of this constructor. Then you don't need the two extra functions that hold the common init code.

wthrowe replied:
Did you intend to do this?  You put a rocket emoji, but I'm not sure what that was supposed to convey.

geoffrey4444 replied:
I needed a symbol to say "I don't plan to do this now, but I'm not saying it's wrong", and there wasn't a good option

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -350,8 +390,86 @@ class BinaryCompactObject : public DomainCreator<3> {
+    using type = double;
+    static constexpr Options::String help = {
+        "Outer boundary or pivot point of the map"};
+    using group = ExpansionMap;
```
For these options that have the group name in the struct name, please add a `name()` function to overwrite it and remove the redundancy in the names. E.g. for this option you can write: `static std::string name() noexcept { return "OuterBoundary"; }`.

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -350,8 +390,86 @@ class BinaryCompactObject : public DomainCreator<3> {
+  /// If the two names are same then a linear radial scaling is used instead of
+  /// a cubic scaling.
+  struct ExpansionFunctionOfTimeNames {
+    using type = std::array<std::string, 2>;
```
I may have asked this before in a different PR, but do these names need to be input options or can the domain creator just choose them appropriately? How would I decide the name when writing an input file?

geoffrey4444 replied:
They need to be options because they must match whatever the user of a spec run chose for them, and this might change from run to run

geoffrey4444 replied:
Once the control system in spectre is done, perhaps the names can be hard coded and these options can be removed, though

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -350,8 +390,86 @@ class BinaryCompactObject : public DomainCreator<3> {
+        "updated."};
+  };
+
+  struct ExpansionMap {
```
Perhaps wrap all time-dependence-related options in another option group named "TimeDependence"?

geoffrey4444 replied:
I prefer the name "TimeDependentMaps" to distinguish the option from the TimeDependence option that lets users add an arbitrary TimeDependence to the domain

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -364,6 +482,20 @@ class BinaryCompactObject : public DomainCreator<3> {
+  using options =
+      tmpl::conditional_t<detail::is_time_dependence_enabled(Metavariables{}),
+                          tmpl::append<time_dependent_options,
+                                       time_independent_options<Metavariables>>,
```
(optional) I think the constructor would be a bit nicer if the time-independent options came before the time-dependent options. I understand some options have default values in the constructor to make it less verbose in the tests, so perhaps just do this:

```cpp
using options = tmpl::append<
  tmpl::list<ObjectA, ObjectB, ...>,
  tmpl::conditional_t<time_dep_enabled, time_dep_options, tmpl::list<>>,
  tmpl::list<UseProjectiveMap, ...>
>;
```

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -389,8 +521,22 @@ class BinaryCompactObject : public DomainCreator<3> {
+      "that the CubicScale map uses to represent its time-dependent parameters "
+      "a(t) and b(t). If the domain is time-dependent, the user must also "
+      "specify the InitialTime and InitialExpirationDeltaT options for the "
+      "FunctionsOfTime that this domain relies on."};
```
I don't think you have to add this description here, since the help text already lists the required options automatically. Extra information for each option you can add to the help text of the particular option.

geoffrey4444 replied:
I replaced this text with text explaining how to enable the time-dependent maps.

nilsvu commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -39,14 +37,43 @@ class FunctionOfTime;
 
+template <size_t Dim>
+struct TimeIndependentMetavariablesWithBoundaryCondition {
+  static constexpr bool enable_time_dependence{false};
```
Here's my suggestion for documenting this option:

```cpp
// Here in the metavariables:
struct Metavariables {

  struct domain_options : tt::ConformsTo<domain::protocols::DomainOptions> {
    static constexpr bool enable_time_dependence{false};
  };

};

// In Domain/Protocols/DomainOptions.hpp (new file)
/*!
 * \brief Compile-time options for creating domains
 *
 * A class conforming to this protocol is placed in the metavariables to choose
 * compile-time options for creating domains. The conforming class must provide
 * the following static member variables:
 *
 * - `bool enable_time_dependence`: Whether or not domains may have time-dependent
 *    maps. When set to `true` domain creators may require additional input-file options
 *    related to the time dependence. When set to `false` the domains are static, e.g.
 *    for elliptic problems.
 *
 * Here's an example for a class conforming to this protocol:
 *
 * \snippet Test_BinaryCompactObject.cpp domain_options_metavars
 */
struct DomainOptions {
  template <typename ConformingType>
  struct test {
    static constexpr bool enable_time_dependence = ConformingType::enable_time_dependence;
  };
};
```

As an example for this pattern you can look at how the `factory_creation` options are documented in the metavariables.

You'll have to write `Metavariables::domain_options::enable_time_dependence` at the point where you retrieve the option in the binary domain creator. I suggest to actually remove the existence check of the option and just require it's there, so there's no "hidden default".

geoffrey4444 replied:
I felt that domain::protocols::DomainOptions would be redundant, so I named this domain::protocols::Metavariables (with the name Metavariables clearly implying that these are compile time options, not run-time options). Happy to change the name if you prefer

nilsvu replied:
Sounds reasonable to me 👍

nilsvu commented on src/Domain/CMakeLists.txt:
```diff
@@ -42,6 +42,7 @@ spectre_target_headers(
   LogicalCoordinates.hpp
   MinimumGridSpacing.hpp
   OptionTags.hpp
+  Protocols.hpp
```
Please place this file in a subdirectory named `Protocols` and name it the same as the struct it defines. We have been moving to this pattern for tags and actions as well, so over time we don't end up with single files that contain long lists of tags/actions/protocols.

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -52,6 +64,24 @@ namespace Frame {
 struct Logical;
 }  // namespace Frame
+
+namespace {
```
Put anonymous namespaces only in cpp files. Here you can name the namespace e.g. `BinaryCompactObject_detail`. Alternatively, you can just delete this helper struct and put the boolean in every executable's metavars. That's a bit annoying to do now, but we should either do it now or in a follow-up PR, e.g. when adding support to `ExportCoordinates`.

geoffrey4444 replied:
I renamed the namespace to `BinaryCompactObject_detail`. I'm not yet convinced that the PR to update ExportCoordinates is the time to add the struct to every executable (or if there might be executables that should not be required to provide this struct in its metavariables), but I'd like to deal with that when I do the next PR, updating ExportCoordinates, not this one

nilsvu commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -358,12 +412,167 @@ class BinaryCompactObject : public DomainCreator<3> {
+  /// the expansion map.
+  ///
+  /// If the two names are same then a linear radial scaling is used instead of
+  /// a cubic scaling.
```
(optional) It's more useful to have details like this in the `help` string than in Doxygen. I'm not sure if options with help strings even need Doxygen documentation.

nilsvu commented on src/Domain/Protocols.hpp:
```diff
@@ -0,0 +1,35 @@
+#pragma once
+
+namespace domain::protocols {
+/*
```
Please add `/*!` so this appears in Doxygen

nilsvu commented on src/Domain/Protocols.hpp:
```diff
@@ -0,0 +1,35 @@
+struct Metavariables {
+  template <typename ConformingType>
+  struct test {
+    using enable_time_dependent_maps_type = const bool;
+    using enable_time_dependent_maps_return_type =
+        decltype(ConformingType::enable_time_dependent_maps);
+    static_assert(std::is_same_v<enable_time_dependent_maps_type,
+                                 enable_time_dependent_maps_return_type>,
+                  "The metavariable 'enable_time_dependent_maps' should be a "
+                  "static constexpr bool.");
```
I think this suffices (or did you get errors?):
```suggestion
    static constexpr bool enable_time_dependent_maps = ConformingType::enable_time_dependent_maps;
```

geoffrey4444 replied:
> I think this suffices (or did you get errors?):
> 
> ```suggestion
> 
>     static constexpr bool enable_time_dependent_maps = ConformingType::enable_time_dependent_maps;
> 
> ```

I got unused variable warnings when I did it your way, and also your way allowed the user to specify any type that converts to bool, not just bool. For both reasons I prefer this way

nilsvu commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -39,14 +38,55 @@ class FunctionOfTime;
 
+template <size_t Dim>
+struct TimeIndependentMetavariablesWithBoundaryCondition {
+  struct domain_metavariables
```
I suggest to name this just `domain`, since the `_metavariables` suffix is redundant.

nilsvu reviewed (changes requested):
I'm happy with squashing things, including these comments.

nilsvu commented on tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:
```diff
@@ -39,14 +38,55 @@ class FunctionOfTime;
 using Excision = domain::creators::BinaryCompactObject::Excision;
 
+template <size_t Dim>
+struct TimeIndependentMetavariablesWithBoundaryCondition {
```
(optional) I think you can just add a `bool TimeDependent` template parameter to this struct, delete the `TimeIndependent` from the name and assign `enable_time_dependent_maps = TimeDependent` below. Then you only need 2 instead of 4 structs. You could even get down to a single struct with another template parameter `UseBoundaryConditions` and `using system = tmpl::conditional_t<...>`.

geoffrey4444 replied:
I replaced this with just one struct

nilsvu reviewed (dismissed):
LGTM

wthrowe commented on src/Domain/Protocols/Metavariables.hpp:
```diff
@@ -0,0 +1,35 @@
+ * \snippet Domain/Test_Protocols.cpp domain_metavariables_example
+ *
+ */
+struct Metavariables {
```
The commit message refers to the old name.

geoffrey4444 replied:
Reworded

wthrowe commented on src/Domain/Protocols/Metavariables.hpp:
```diff
@@ -0,0 +1,35 @@
+    using enable_time_dependent_maps_type = const bool;
+    using enable_time_dependent_maps_return_type =
+        decltype(ConformingType::enable_time_dependent_maps);
+    static_assert(std::is_same_v<enable_time_dependent_maps_type,
```
`#include <type_traits>`

wthrowe commented on tests/Unit/Domain/Test_Protocols.cpp:
```diff
@@ -0,0 +1,19 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
```
Unused.

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -51,6 +64,23 @@ namespace Frame {
+// `domain_parameters::enable_time_dependent_maps` is `true`, then
+// inherit from `std::true_type`; otherwise, inherit from `std::false_type`.
+template <typename Metavariables, typename = std::void_t<>>
+struct enable_time_dependent_maps : std::false_type {};
```
`#include <type_traits>`

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -19,7 +20,9 @@
 #include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/NoSuchType.hpp"
 #include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/CreateHasStaticMemberVariable.hpp"
```
Unused.

geoffrey4444 replied:
Thanks for catching this! When I changed to using a protocol, I forgot to remove the headers I added for handling a constexpr static member variable

wthrowe commented on src/Domain/Creators/BinaryCompactObject.hpp:
```diff
@@ -19,7 +20,9 @@
 #include "Options/Auto.hpp"
 #include "Options/Options.hpp"
 #include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/NoSuchType.hpp"
```
Unused.

geoffrey4444 commented:
Thanks for those last-minute catches, @wthrowe ! I just pushed fixup commits

wthrowe commented:
Squash.

geoffrey4444 commented:
Squashed, @wthrowe

nilsvu commented:
@nikwit do you want to take another look before we merge?

nikwit commented:
please go ahead and merge!
